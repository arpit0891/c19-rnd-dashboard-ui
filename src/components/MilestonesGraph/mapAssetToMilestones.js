import moment from 'moment'
import { pipe } from 'sanctuary'
import { status, phasesInOrder, timelinesEstimates } from './constants'
import { mapTwoAtTime } from 'utils/utils'

const { skipped } = status

const filterAndSortMilestones = milestones =>
  milestones
    .filter(({ status }) => status !== skipped)
    .sort(
      ({ name: nameA }, { name: nameB }) =>
        phasesInOrder.indexOf(nameA) - phasesInOrder.indexOf(nameB)
    )

const addDurationToMilestones = ({ now, delta }) => milestones => {
  let buffer = []
  for (let i = 1; i <= milestones.length; i++) {
    const first = milestones[i - 1]
    if (first.duration) {
      continue
    }
    const second = milestones[i] ? milestones[i] : { date: now }
    const { date: lastDate } = first
    const { date: newDate } = second
    const start = moment(lastDate)
    const end = moment(newDate)
    const duration = end.diff(start, 'days') || delta[first.name]
    buffer = buffer.concat({ ...first, duration, start, end })
  }
  return buffer
}

const addPercentageToMilestones = ms => {
  const totals = ms.map(({ values }) =>
    values.reduce((acc, { duration = 0 }) => duration + acc, 0)
  )
  const [max] = totals.slice().sort((a, b) => b - a)
  return ms.map(m => ({
    ...m,
    values: m.values.map(x => ({
      ...x,
      percentage: Math.round(((x.duration * 100) / max) * 100) / 100,
    })),
  }))
}

const mapResultToMilestoneStructure = ms =>
  ms.map(m => ({
    ...m,
    values: m.values.map(({ name, percentage, start, end }) => ({
      className: name,
      value: `${percentage}%`,
      start: moment(start).toISOString(),
      end: moment(end).toISOString(),
    })),
  }))

const getAdditionalPhases = phase => {
  if (!phase) {
    return phasesInOrder
  }
  const index = phasesInOrder.indexOf(phase)
  return phasesInOrder.filter(p => phasesInOrder.indexOf(p) > index)
}

const estimateFutureDates = (ms, delta) => {
  return mapTwoAtTime((first, second) => {
    const duration = delta[first.name]
    const newEnd = moment(first.start).add(duration, 'days').toISOString()
    const newFirst = { ...first, end: newEnd, duration }
    const newSecond = { ...second, start: newEnd }
    return second ? [newFirst, newSecond] : [newFirst]
  }, ms)
}

// const getLatestsKnownDate = (ms) => {
//   const [lastButOne, last] = ms.slice(-2)
//   const {date, start, end} = last

// }

const transformWithDurations = ({ now, delta }) =>
  pipe([filterAndSortMilestones, addDurationToMilestones({ now, delta })])

const t2 = pipe([addPercentageToMilestones, mapResultToMilestoneStructure])

export const mapAssetToMilestones = now => ({ milestones }) => {
  if (!milestones) {
    return []
  }

  const actualMilestonesWithDuration = transformWithDurations({
    now,
    delta: timelinesEstimates.actual,
  })(milestones)
  const [latestKnownDate] = actualMilestonesWithDuration.slice(-1)
  const estimationPhases = getAdditionalPhases(
    latestKnownDate.name
  ).map(phase => ({ name: phase }))
  const optimisticEstimations = estimateFutureDates(
    [latestKnownDate].concat(estimationPhases),
    timelinesEstimates.optimistic
  )
  const pesimisticEstimations = estimateFutureDates(
    [latestKnownDate].concat(estimationPhases),
    timelinesEstimates.pesimistic
  )
  const result = [
    {
      name: 'Optimistic',
      values: actualMilestonesWithDuration.concat(optimisticEstimations),
    },
    {
      name: 'Pesimistic',
      values: actualMilestonesWithDuration.concat(pesimisticEstimations),
    },
    {
      name: 'Actual',
      values: actualMilestonesWithDuration,
    },
  ]
  return t2(result)
}
