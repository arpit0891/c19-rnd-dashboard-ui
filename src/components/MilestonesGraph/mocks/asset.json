{
  "brandName": "",
  "chemicalName": "mrna-1273",
  "conditionOrDisease": "covid-19",
  "countries": "united states",
  "countryCodes": "usa",
  "currentStatus": "phase 1",
  "indication": "covid-19",
  "interventionType": "vaccine - prophylactic",
  "milestones": [
    {
      "category": "pre-clinical",
      "date": null,
      "milestoneId": 12,
      "name": "pre_clinical_studies",
      "status": "SKIPPED"
    },
    {
      "category": "pre-clinical",
      "date": "Mon, 13 Jan 2020 00:00:00 GMT",
      "milestoneId": 13,
      "name": "lead_selection",
      "status": "COMPLETED"
    },
    {
      "category": "manufacturing",
      "date": "Fri, 07 Feb 2020 00:00:00 GMT",
      "milestoneId": 21,
      "name": "clinical_batch",
      "status": "COMPLETED"
    },
    {
      "category": "regulatory",
      "date": "Wed, 04 Mar 2020 00:00:00 GMT",
      "milestoneId": 31,
      "name": "ind",
      "status": "COMPLETED"
    },
    {
      "category": "clinical_development",
      "date": "Mon, 16 Mar 2020 00:00:00 GMT",
      "milestoneId": 41,
      "name": "phase_1",
      "status": "COMPLETED"
    },
    {
      "category": "pre-clinical",
      "date": "Sat, 11 Jan 2020 00:00:00 GMT",
      "milestoneId": 11,
      "name": "discovery",
      "status": "COMPLETED"
    }
  ],
  "moleculeType": "nucleic acid based therapies/vaccines",
  "notes": "mrna-based vaccine",
  "numSites": "2",
  "otherPartners": "cepi",
  "phase": "1",
  "preferredName": "mrna-1273",
  "productId": 1,
  "productType": "",
  "repurposed": "new",
  "siteLocations": [
    {
      "city": "Seattle",
      "country": "US",
      "lat": 47.6169397,
      "lng": -122.329572,
      "locationId": "a6018262eeb2425434539c72bd2bc1cf527bca54",
      "name": "1730 Minor Ave, Seattle, WA 98101, USA",
      "state": "WA"
    },
    {
      "city": "Atlanta",
      "country": "US",
      "lat": 33.7934917,
      "lng": -84.3195325,
      "locationId": "a20a7fbfc641e0fed17496f7f38d14ea4ba6c6a3",
      "name": "1405 E Clifton Rd NE, Atlanta, GA 30322, USA",
      "state": "GA"
    }
  ],
  "sources": [
    "https://www.nih.gov/news-events/news-releases/nih-clinical-trial-investigational-vaccine-covid-19-begins",
    "https://www.modernatx.com/modernas-work-potential-vaccine-against-covid-19",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=2",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=1",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=3",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=4",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=5",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=6",
    "https://clinicaltrials.gov/ct2/show/nct04283461?term=coronavirus+vaccine&draw=7",
    "https://corona.kpwashingtonresearch.org/"
  ],
  "sponsors": [
    {
      "sponsorId": "375ad6c4b12a03acefcf5e9b052423279351910a",
      "sponsorName": "Moderna"
    },
    {
      "sponsorId": "69301ed353817a1eb80b77b32f82e5809965a871",
      "sponsorName": "National Institute of Allergy and Infectious Diseases"
    }
  ],
  "status": "ongoing",
  "therapeuticApproach": "unknown",
  "trialId": "nct04283461"
}
