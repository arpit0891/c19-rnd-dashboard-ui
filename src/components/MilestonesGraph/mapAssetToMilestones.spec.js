import asset from './mocks/asset.json'
import { mapAssetToMilestones } from './mapAssetToMilestones'
import moment from 'moment'

it('should return the expected result', () => {
  const now = moment('2020-05-09T20:47:50.756Z')
  expect(mapAssetToMilestones(now)(asset)).toMatchSnapshot()
})
