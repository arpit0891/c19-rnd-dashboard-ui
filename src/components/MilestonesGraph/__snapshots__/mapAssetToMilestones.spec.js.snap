// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the expected result 1`] = `
Array [
  Object {
    "name": "Optimistic",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.17%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.13%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.21%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.02%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-09T20:47:50.756Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "4.6%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-30T01:00:00.000Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "6.39%",
      },
      Object {
        "className": "phase_2",
        "end": "2020-10-27T00:00:00.000Z",
        "start": "2020-05-30T01:00:00.000Z",
        "value": "12.78%",
      },
      Object {
        "className": "phase_3",
        "end": "2021-06-24T01:00:00.000Z",
        "start": "2020-10-27T00:00:00.000Z",
        "value": "20.44%",
      },
      Object {
        "className": "nda",
        "end": "2021-07-15T01:00:00.000Z",
        "start": "2021-06-24T01:00:00.000Z",
        "value": "1.79%",
      },
      Object {
        "className": "on_market",
        "end": "2021-08-14T01:00:00.000Z",
        "start": "2021-07-15T01:00:00.000Z",
        "value": "2.56%",
      },
    ],
  },
  Object {
    "name": "Pesimistic",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.17%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.13%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.21%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.02%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-09T20:47:50.756Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "4.6%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-09-12T00:00:00.000Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "15.33%",
      },
      Object {
        "className": "phase_2",
        "end": "2021-06-09T01:00:00.000Z",
        "start": "2020-09-12T00:00:00.000Z",
        "value": "23%",
      },
      Object {
        "className": "phase_3",
        "end": "2022-06-09T01:00:00.000Z",
        "start": "2021-06-09T01:00:00.000Z",
        "value": "31.09%",
      },
      Object {
        "className": "nda",
        "end": "2022-08-08T01:00:00.000Z",
        "start": "2022-06-09T01:00:00.000Z",
        "value": "5.11%",
      },
      Object {
        "className": "on_market",
        "end": "2023-02-04T00:00:00.000Z",
        "start": "2022-08-08T01:00:00.000Z",
        "value": "15.33%",
      },
    ],
  },
  Object {
    "name": "Actual",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.17%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.13%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.21%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.02%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-09T20:47:50.756Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "4.6%",
      },
    ],
  },
]
`;
